fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    fetch('/api/todos', {
+      method: 'GET'
+    })
+      .then(res => res.json())
+      .then(data => {
+        const arrayTodos = [];
+        for (let i = 0; i < data.length; i++) {
+          arrayTodos.push(data[i]);
+        }
+        this.setState({ todos: arrayTodos });
+      });
+    /**
+* Use fetch to send a GET request to `/api/todos`.
+* Then 😉, once the response JSON is received and parsed,
+* update state with the received todos.
+*/
+  }
+
+  addTodo(newTodo) {
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => res.json())
+      .then(data => {
+        const oldArray = this.state.todos;
+        const newToDo = data;
+        const newArray = oldArray.concat(newToDo);
+        this.setState({ todos: newArray });
+      });
+    /**
+* Use fetch to send a POST request to `/api/todos`.
+* Then 😉, once the response JSON is received and parsed,
+* add the created todo to the state array.
+*
+* Do not mutate the original state array, nor any objects within it.
+* https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+*
+* TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+* and specify the "Content-Type" header as "application/json"
+*
+* TIP: Use Array.prototype.concat to create a new array containing the contents
+* of the old array, plus the object returned by the server.
+*/
+  }
+
+  toggleCompleted(todoId) {
+    const todosArray = this.state.todos;
+    for (let i = 0; i < todosArray.length; i++) {
+      if (todoId === todosArray[i].todoId) {
+        const todo = todosArray[i];
+        const oldStatus = todo.isCompleted;
+        const newStatus = !oldStatus;
+        const editedToDo = todo;
+        editedToDo.isCompleted = newStatus;
```

Comment:
In React, we don't modify objects from state directly. Instead we make a copy. Actually, check the pseudo code, it says to make a new object and send that to the server.